text/0000-add-limits-trait.md

```diff
@@ -0,0 +1,133 @@
+- Feature Name: Limits trait for the rust types
+- Start Date: 2017-12-20
+- RFC PR: 
+- Rust Issue: 
+
+# Summary
+This is an RFC to add a universal trait for the type limits.
+
+# Motivation
+The motivation is quite simple: [make an ability to accept template types with with limits by requiring a trait](https://stackoverflow.com/questions/47904954/rust-finding-the-maximum-allowable-value-for-generic-data-type-t) 
+so that it simplifies and generalizes the code. Another motivation is that we have all that `max_value()` and `min_value()` implemented as
+usual methods of a type implementation, generalizing this to a trait makes the code simplier and avoids duplicating code. Also, looking at
+C++ template of `std::numeric_limits` tells us we must have this thing too because it is easier to use.
+
+# Detailed design
+The design is quite simple: put everything related to the limits what can be generalized into a separate trait in the standard library:
+
+```rust
+trait Limits<T> {
+    fn min_value() -> T;
+    fn max_value() -> T;
+}
+
+#[derive(Debug, Copy, Clone)]
+struct A {
+    field: u64,
+}
+impl Limits<A> for A {
+    fn min_value() -> A {
+        A {
+            field: 0u64,
+        }
+    }
+
+    fn max_value() -> A {
+        A {
+            field: 5u64,
+        }
+    }
+}
+impl Limits<u32> for u32 {
+    fn min_value() -> u32 {
+        0
+    }
+    fn max_value() -> u32 {
+        10u32
+    }
+}
+
+fn get_limits<T: Limits<T> + std::fmt::Debug>(_t: T) {
+    println!("Minimum value: {:?}", T::min_value());
+    println!("Maximum value: {:?}", T::max_value());
+}
+
+fn main() {
+    let a = A { field: 6u64 };
+    let num = 10u32;
+    get_limits(a);
+    get_limits(num);
+}
+
+```
+
+Here we have a generalized function `get_limits` which accepts its argument with requirement for trait `Limits` implementation. As long
+as a type implements this trait, this function will succeed and will produce expected results. It's worth mentioning that a type can implement
+different limits type, not only for itself: `struct A` can have both `Limits<A>` and `Limits<u32>` implementations: we may simply add another
```

Comment:
What is the use case for `Limits<A> for B` where `A != B` ?

Reply:
1. A "shorthand" for inner type of strong types: `struct Id(pub u64)` and `impl Limits<u64> for Id`. But this is arguably since we made strong type exactly for hiding the inner type and we will violate the strong type's meaning by exposing its internal type in a trait.
2. Well, I thought I had a reason for this, but I don't remember it. Could you think any reason?

Reply:
On 2. I'm afraid not =)

Reply:
So as I also don't remember why I did it so (what I was thinking about during writing), we may turn it into a simple trait then.